### ISSUE TITLE
KSP2 crash when using @JsonQualifier

### ISSUE THREAD

**Issue by Jean-Daniel** [square/moshi#1874]
When trying to generate Adapter for classes using a @JsonQualifier, KSP fails with the error:

```

> Task :app:kspDebugKotlin FAILED
e: [ksp] Error preparing ApiUserItem: jdk.proxy5/jdk.proxy5.$Proxy154.value(Unknown Source)
com.squareup.moshi.kotlin.codegen.ksp.MoshiApiUtilKt.generator(MoshiApiUtil.kt:77)
com.squareup.moshi.kotlin.codegen.ksp.JsonClassSymbolProcessor.adapterGenerator(JsonClassSymbolProcessorProvider.kt:122)
com.squareup.moshi.kotlin.codegen.ksp.JsonClassSymbolProcessor.process(JsonClassSymbolProcessorProvider.kt:92)
com.google.devtools.ksp.impl.KotlinSymbolProcessing.execute(KotlinSymbolProcessing.kt:538)
com.google.devtools.ksp.impl.KSPLoader$Companion.loadAndRunKSP(KSPLoader.kt:36)
com.google.devtools.ksp.impl.KSPLoader.loadAndRunKSP(KSPLoader.kt)
…
```

The class being compiled is simply:

```kotlin
@Retention(AnnotationRetention.RUNTIME)
@JsonQualifier
annotation class ISO8061

@JsonClass(generateAdapter = true)
open class ApiUserItem(
  @field:Json(name = "slug") val slug: String,
  @field:Json(name = "date_added") @field:ISO8061 val date: Long,
)
```

You can use this reduced test case to reproduce the bug.

```
git clone https://github.com/Jean-Daniel/moshiksp2
cd moshiksp2
./gradlew :app:assembleDebug
> Task :app:kspDebugKotlin FAILED
…
```

**Comment by ZacSweers** [comment]
Upstream KSP2 bug, please file with them

**Comment by ting-yuan** [comment]
This is caused by a [breaking change](https://github.com/google/ksp/blob/main/docs/ksp2api.md#evaluation-of-enum-entries-in-annotation-arguments) in KSP2. @ZacSweers could [this copy](https://github.com/square/moshi/blob/d17b6736740b49dd167dc0ca32c1dcef5ebb48f3/moshi-kotlin-codegen/src/main/java/com/squareup/moshi/kotlin/codegen/ksp/shadedUtil.kt) be deleted and let moshi use the implementation from KSP?

**Comment by ting-yuan** [comment]
Removing it doesn't seem to break any tests:
https://github.com/square/moshi/pull/1875

**Comment by Jean-Daniel** [comment]
Can this bug be reopened as this is clearly a Moshi bug as it is the moshi `shadedUtil.kt` version that is faulty.

**Comment by ZacSweers** [comment]
The linked fix isn't a useable fix because it's an unstable API. A more appropriate one would be to update the copied impl to the new one

**Comment by malikmudassarhussain** [comment]
While migrating to Koline 2.x I had to fallback to moshi-ksp from moshi-ir because of [this](https://issuetracker.google.com/issues/381407900) IR issue and now I am stuck on this error. Is there any quick fix as the IR one will be fixed in Kotlin 2.1.20

**Comment by Jean-Daniel** [comment]
> Is there any quick fix as the IR one will be fixed in Kotlin 2.1.20

Yes, just keep using KSP 1 instead of KSP 2, as only KSP 2 is impacted by this issue.

**Comment by Jean-Daniel** [comment]
> The linked fix isn't a useable fix because it's an unstable API. A more appropriate one would be to update the copied impl to the new one

I'm not sure I get the 'unstable API' argument.

Using an unstable API can cause mochi to stop working when updating KSP, but this is the same as copying the file, as updating KSP to a version with a different `shadedUtil.kt` file/API is also breaking Moshi, as shown here.

**Comment by ygnessin** [comment]
> Yes, just keep using KSP 1 instead of KSP 2, as only KSP 2 is impacted by this issue.

KSP1 [does not support Kotlin 2.1](https://github.com/google/ksp/blob/main/docs/ksp2.md#ksp1-deprecation-schedule), which has now been released for 4 months. This issue is blocking my team from updating our project to Kotlin 2.1.

@ZacSweers Is there any possibility of re-considering a fix for this? Thank you!!!

**Comment by SteinerOk** [comment]
@JakeWharton, @ZacSweers Have the same problem with JsonQualifier in stable KSP 2.1.20-2.0.0

**Comment by masztalski** [comment]
I have checked and removing @JsonQualifiers with KSP 2.1.21-2,0.2 resolves an issue.
However it is not way to go as I need them.

**Comment by ygnessin** [comment]
@ZacSweers Apologies for the bump, but any updates on this? Kotlin 2.2 has now been released for over a month, and even though we have the `ksp.useKSP2=false` workaround it would be preferable to migrate to KSP2.

I see https://github.com/square/moshi/pull/1966 has been opened as a potential solution, but it seems to have stagnated. Any idea on an expected timeline?

Thank you so much!

**Comment by arcao** [comment]
When you plan to release version with KSP2 support?

AGP 8.13.0 started to warn about deprecated KSP1:
> We noticed you are using KSP1 which is deprecated and support for it will be removed soon - please migrate to KSP2 as soon as possible. KSP1 will no longer be compatible with Android Gradle Plugin 9.0.0 (and above) and Kotlin 2.3.0 (and above)

**Comment by LukasPrediger** [comment]
This issue persists with the current version of KSP and Moshi. It is not fixed!

**Comment by SteinerOk** [comment]
Hello @JakeWharton, @ZacSweers, do you have any estimated timeframe for releasing a fix for this issue, or should I migrate to other serialization library if want to use Kotlin 2.3.0 + KSP 2.3.4?
Thanks :)

**Comment by tanishq-meesho** [comment]
Do we have any update here ? 
@JakeWharton

**Comment by BenderNK** [comment]
Jake wrote [on another issue](https://github.com/square/moshi/issues/2079#issuecomment-4051193660) that he personally won't be maintaining Moshi anymore - essentially because Moshi was optimized for Java and there is little reason to use Moshi for Kotlin. Once we discovered that post from him, our team migrated away from Moshi. Good luck.
